### ISSUE TITLE
Mocking/stubbing/spying not working right in Quarkus 3.x

### ISSUE THREAD

**Issue by edeandrea** [quarkusio/quarkus#32300]
### Describe the bug

This was created off of #31400.

In the [`rest-fights` application from Quarkus Superheroes](https://github.com/edeandrea/quarkus-super-heroes/tree/quarkus3/rest-fights) there's some issue with mocking/stubbing/spying within tests. See https://github.com/edeandrea/quarkus3-supes-mocking/actions/runs/4574816927/jobs/8076900165 for full the full test execution, failures, and errors.

See https://github.com/edeandrea/quarkus3-supes-mocking/blob/main/README.md for full details.

### Expected behavior

I would expect existing tests that use mocking/stubbing/spying to work correctly.

### Actual behavior

Tests that use mocking/stubbing/verification do not work properly.

### How to Reproduce?

I created https://github.com/edeandrea/quarkus3-supes-mocking which is as stripped down as I can get it (no Kafka/stork/rest client/rest endpoints/mapstruct/removing timeout fault tolerance/etc) to still show the problem. The repo has 2 identical versions of the stripped down app, 1 with Quarkus 2 & 1 with Quarkus 3. All tests pass in the quarkus 2 version yet there are failures in the Quarkus 3 version.

### Output of `uname -a` or `ver`

`Darwin edeandrea-m1pro 22.3.0 Darwin Kernel Version 22.3.0: Mon Jan 30 20:38:37 PST 2023; root:xnu-8792.81.3~2/RELEASE_ARM64_T6000 arm64`

### Output of `java -version`

```
openjdk version "17.0.6" 2023-01-17
OpenJDK Runtime Environment Temurin-17.0.6+10 (build 17.0.6+10)
OpenJDK 64-Bit Server VM Temurin-17.0.6+10 (build 17.0.6+10, mixed mode)
```

### GraalVM version (if different from Java)

```
openjdk 17.0.6 2023-01-17
OpenJDK Runtime Environment GraalVM CE 22.3.1 (build 17.0.6+10-jvmci-22.3-b13)
OpenJDK 64-Bit Server VM GraalVM CE 22.3.1 (build 17.0.6+10-jvmci-22.3-b13, mixed mode, sharing)
```

### Quarkus version or git rev

3.0.0.CR1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

```
Apache Maven 3.8.4 (9b656c72d54e5bacbed989b64718c159fe39b537)
Maven home: /Users/edeandre/.m2/wrapper/dists/apache-maven-3.8.4-bin/52ccbt68d252mdldqsfsn03jlf/apache-maven-3.8.4
Java version: 17.0.6, vendor: Eclipse Adoptium, runtime: /Users/edeandre/.sdkman/candidates/java/17.0.6-tem
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "13.2.1", arch: "aarch64", family: "mac"
```

**Comment by geoand** [comment]
Thanks a lot @edeandrea, I'll check it out

**Comment by edeandrea** [comment]
> So the error I am seeing is:
> 
> ```posh
> [ERROR] Tests run: 26, Failures: 10, Errors: 0, Skipped: 0, Time elapsed: 13.028 s <<< FAILURE! - in io.quarkus.sample.superheroes.fight.service.FightServiceTests
> [ERROR] io.quarkus.sample.superheroes.fight.service.FightServiceTests.findRandomFightersNoneFound  Time elapsed: 0.039 s  <<< FAILURE!
> Wanted but not invoked:
> fightService_Subclass.findRandomHero();
> -> at io.quarkus.sample.superheroes.fight.service.FightServiceTests.findRandomFightersNoneFound(FightServiceTests.java:218)
> 
> However, there was exactly 1 interaction with this mock:
> fightService_Subclass.findRandomFighters();
> -> at io.quarkus.sample.superheroes.fight.service.FightServiceTests.findRandomFightersNoneFound(FightServiceTests.java:205)
> ```

That is correct. The mocking/spying is verifying that certain methods were called, which you can clearly see in the test run log, yet the verification is failing saying that they weren't called.

**Comment by Ladicek** [comment]
I admit I'm totally lost in Mockito's generated subclass, which extends ArC's generated subclass, which extends the application class, and it's all reactive... I'll use the big hammer and do `git bisect` first :-)

**Comment by Ladicek** [comment]
The issue seems to have been introduced between 3.0.0.Alpha3 and Alpha4. I'll look deeper.

**Comment by Ladicek** [comment]
This is the offending commit: 0ed5b7b4a7b1102e90757b38e0f4b402173e26ed. Reverting it makes all the tests pass. I'll look deeper.

**Comment by Ladicek** [comment]
What I found is:

1. The moment we instantiate a bean with intercepted method, we instantiate an ArC `_Subclass`. That subclass will, in turn, instantiate `InterceptedMethodMetadata` for all intercepted methods, which includes remembering `this` as the instance on which the method should be called after all interceptors got a chance.
2. When the test calls a method on a mocked bean, it will actually be called on an instance of a Mockito-generated subclass of ArC's `_Subclass`. The `InvocationContext.target` even correctly points to that instance. However, `InvocationContext.proceed()` totally ignores the `target` and will proceed to use the instance remembered in `InterceptedMethodMetadata`.

I believe that is a bug, and should be reasonably straightforward to fix. I'll continue tomorrow.

**Comment by Ladicek** [comment]
Here's a PR that fixes this: #32405

**Comment by edeandrea** [comment]
Thank you all for working on this quickly! Once CR2 is out I will re-test and reopen this issue if the problem seems to persist.
